discord_oauth.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+func (r *Router) DiscordSignin(w http.ResponseWriter, req *http.Request) {
+	state := generateStateOauthCookie(w)
+	redirectUri := fmt.Sprintf("%s/discord/callback", r.rootURL)
+	url := fmt.Sprintf("https://discord.com/oauth2/authorize?client_id=%s&response_type=code&redirect_uri=%v&scope=identify+guilds.join&state=%v", r.bot.ClientId, url.QueryEscape(redirectUri), state)
+	http.Redirect(w, req, url, http.StatusTemporaryRedirect)
+}
+
+func (r *Router) DiscordCallback(w http.ResponseWriter, req *http.Request) {
+	stateCookie, err := req.Cookie("oauthstate")
+	if err != nil {
+		http.Error(w, "Missing oauthstate cookie", http.StatusBadRequest)
+		return
+	}
+	stateParam := req.URL.Query().Get("state")
+	if stateParam == "" {
+		http.Error(w, "Missing state url parameter", http.StatusBadRequest)
+		return
+	}
+	if stateCookie.Value != stateParam {
+		http.Error(w, "State cookie and state parameter don't match", http.StatusBadRequest)
+		return
+	}
+
+	code := req.URL.Query().Get("code")
+	authToken, err := getDiscordAuthToken(r.rootURL, r.bot, code)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Error getting discord auth token", http.StatusForbidden)
+		return
+	}
+
+	discordUser, err := getDiscordUser(authToken)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Error getting user information", http.StatusForbidden)
+		return
+	}
+
+	userId, hasUserId := getUserIDFromContext(req.Context())
```

Comment:
This route is under the `EnforceJwtMiddleware` so it actually doesn't need to do this error check